triton-vm/src/stark.rs

```diff
@@ -1144,128 +1032,212 @@ impl Stark {
+        let num_input_rows = quotient_multicoset_evaluations.nrows();
+        let num_cosets = quotient_multicoset_evaluations.ncols();
+        let num_output_rows = num_input_rows * num_cosets / NUM_SEGMENTS;
+
+        assert!(num_input_rows.is_power_of_two());
+        assert!(num_cosets.is_power_of_two());
+        assert!(num_output_rows.is_power_of_two());
+        assert!(NUM_SEGMENTS.is_power_of_two());
         assert!(
-            num_rows > num_segments,
-            "trace domain length: {num_rows} versus num segments: {num_segments}",
+            num_input_rows >= num_cosets,
+            "working domain length: {num_input_rows} versus num cosets: {num_cosets}",
+        );
+        assert!(
+            num_cosets >= NUM_SEGMENTS,
+            "num cosets: {num_cosets} versus num segments: {NUM_SEGMENTS}",
         );
 
-        // Matrix `quotients` contains q(Ψ · ι^j · ω^i) in location (i,j) where ω is the
-        // trace domain generator, and where iota is an Fth root of ω such that ι^F = ω,
-        // where F is `num_quotient_segments`. So `quotients` contains q(Ψ · ι^(j+i·F)).
-
-        // We need F-tuples from this matrix of elements separated by N/F rows.
-        let step_size = num_rows / num_segments;
-        let quotient_segments = (0..num_rows)
+        // Re-arrange data in preparation for iNTT:
+        // Move appropriate powers of ξ^(k·l) with the same k into the same row. Change
+        // the matrix's dimensions from N×M to L×K, with row majority (“`C`”) to get
+        // contiguous row slices for iNTT.
+        let mut quotient_segments = ndarray_helper::par_zeros((num_output_rows, NUM_SEGMENTS));
+        quotient_segments
+            .axis_iter_mut(Axis(0))
             .into_par_iter()
-            .flat_map(|jif| {
-                let col_idx = jif % num_segments;
-                let start_row = (jif - col_idx) / num_segments;
-                quotient_multicoset_evaluations
-                    .slice(s![start_row..; step_size, col_idx])
-                    .to_vec()
-            })
-            .collect();
-        let mut quotient_segments =
-            Array2::from_shape_vec((num_rows, num_segments), quotient_segments).unwrap();
-
-        // Matrix `quotient_segments` now contains q(Ψ · ι^(j+i·F+l·N/F)) in cell
-        // (j+i·F, l). So *row* j+i·F contains {q(Ψ · ι^(j+i·F+l·N/F)) for l in [0..F-1]}.
+            .enumerate()
+            .for_each(|(output_row_idx, mut output_row)| {
+                for (output_col_idx, cell) in output_row.iter_mut().enumerate() {
+                    let exponent_of_iota = output_row_idx + output_col_idx * num_output_rows;
+                    let input_row_idx = exponent_of_iota / num_cosets;
+                    let input_col_idx = exponent_of_iota % num_cosets;
+                    *cell = quotient_multicoset_evaluations[[input_row_idx, input_col_idx]];
+                }
+            });
```

Comment:
This code is certainly different from the code before, which used slicing. I realize that slicing might not be possible in the general case. Nevertheless, I wonder if this change incurs a performance penalty when it is possible.

Reply:
My benchmark (fibonacci 10_000, n = 11) indicates that the current code is marginally faster. `segmentify` clocks in at 2.20% of the compute time (current state of the PR) vs 2.53% of the compute time (slicing). It is quite possible that the dominant factor in this comparison is noise.

Reply:
Well I did not expect that. I guess `ndarray`'s slicing also incurs a significant arithmetic overhead. I don't know what goes on under the hood.

Reply:
More benchmarks!

```txt
slicing 1024x8          time:   [74.165 µs 74.958 µs 75.928 µs]
indexing 1024x8         time:   [39.174 µs 39.513 µs 39.921 µs]

slicing 2048x8          time:   [145.72 µs 146.13 µs 146.67 µs]
indexing 2048x8         time:   [50.944 µs 51.450 µs 52.138 µs]

slicing 4096x8          time:   [301.74 µs 304.43 µs 307.94 µs]
indexing 4096x8         time:   [71.979 µs 72.387 µs 72.824 µs]

slicing 8192x8          time:   [599.24 µs 601.04 µs 603.16 µs]
indexing 8192x8         time:   [115.55 µs 117.61 µs 120.18 µs]
```

I'd say we have a clear winner. :relieved: